Add dist config docs (#194)
Oneflow-Inc/libai: docs/source/tutorials/Distributed Configuration.md

# Distributed Configuration

> In LiBai, you can try out different parallel modes easily by updating the [training config file](https://github.com/Oneflow-Inc/libai/blob/main/configs/common/train.py).

Reviewer: ```suggestion
In LiBai, you can try out different parallel modes easily by updating the distributed config in [training config file](https://github.com/Oneflow-Inc/libai/blob/main/configs/common/train.py).
```

Author: 已修改~

```python
# Distributed arguments
dist=dict(
        data_parallel_size=1,
        tensor_parallel_size=1,
        pipeline_parallel_size=1,
)
```
For example, you can set `data_parallel_size=2` which will automatically split the input data into two groups for data parallel training.

## Distributed Setting Example

> Here we provide simple examples for users to understand the basic configuration of LiBai's distributed settings. LiBai's **BERT** model supports three parallelism techniques, and here we use 1 node with 8 GPUs as example, if users do not change any default settings, LiBai will execute **data parallel training as default**. You can try out different combinations of parallelism training techniques by updating [bert config file](../../../configs/bert_large_pretrain.py) as follows:

Reviewer: ```suggestion
Here we provide simple examples for users to understand the basic configuration of LiBai's distributed settings. LiBai's **BERT** model supports three parallelism techniques, and here we use 1 node with 8 GPUs as an example. If you do not change any default settings, LiBai will execute **data parallel training as default**. You can try out different combinations of parallelism training techniques by updating [bert config file](../../../configs/bert_large_pretrain.py) as follows:
```

Author: 已修改~

> Here we provide simple examples for users to understand the basic configuration of LiBai's distributed settings. LiBai's **BERT** model supports three parallelism techniques, and here we use 1 node with 8 GPUs as an example. If you do not change any default settings, LiBai will execute **data parallel training as default**. You can try out different combinations of parallelism training techniques by updating [bert config file](../../../configs/bert_large_pretrain.py) as follows:

Reviewer 2: ```suggestion
Here we provide simple examples for users to understand the basic configuration of LiBai's distributed settings. LiBai's **BERT** model supports three parallelism techniques, and here we use 1 node with 8 GPUs as an example. If you do not change any default settings, LiBai will execute **data parallel training** by default. You can try out different combinations of parallelism training techniques by updating [bert config file](../../../configs/bert_large_pretrain.py) as follows:
```

#### **Pure Data Parallel Training on 8 GPUs**

> In this example, the input data will be splitted into 8 parts on batch dim for data parallel training on 8 GPUs.

Reviewer 2: 这个切其实只是逻辑上的，实际并没有切数据，而且熟悉 torch ddp 的用户，应该也知道，分布式数据并行的时候，每个rank直接读取本rank对应的 local batch 数据，并没有切batch的行为，感觉这里其实有可能会有误解

Author: 那我改一下表述，我不太清楚如何描述比较贴切

Reviewer 2: 其实我觉得不需要解释，就是给代码就行了，因为标题已经解释了

> In this example, the model is replicated on 8 GPUs, and each replica handles a part of the input data.

Reviewer 2: ```suggestion
In this example, the model is replicated on 8 GPUs, and each replica handles only a part of the whole dataset.
```

Reviewer 2: 这样如何

Author: 我觉得行

Author: 稍微改了一下，看看最新的版本呢~

```python
from .common.train import train
...

train.dist.data_parallel_size = 8
```

#### **Pure Tensor Parallel Training on 8 GPUs**

In this example, the weight of the layers in the model be splitted into 8 parts for tensor parallel training on 8 GPUs.
```python
from .common.train import train
...

train.dist.tensor_parallel_size = 8
```

#### **Pure Pipeline Parallel Training on 8 GPUs**

> In this example, 8 GPUs will be splitted into 8 stages, different layers of the model will be put on different stages automatically for pipeline parallel training.

Reviewer 2: 其实只有内置的模型才是会自动把模型分到不同的卡上，如果是自己写新的模型需要给模型每个模块去设置 `layer_idx` 才会分，不然就是都堆到 0 号卡上了？

Author: 奥你说的对。。这里要加上例子才行，需要配置一下

Author: 那问题复杂了。。得告诉用户如何配置，等我加一个`Write Models`的例子吧

Reviewer 2: 是的，不然用户真以为这里配置个流水并行，模型就能自动流水了

Reviewer: 这里只是教用户如何对内置模型配置流水并行，对于新的模型怎么增加流水并行的支持，我们在别的地方教他

Author: 或者我在这里声明一下，针对已经配置过流水并行的模型，可以通过这样的设置去跑不同阶段的流水，然后再refer一下`Write Models`文档，说如果不知道如何配置流水并行的模型，可以参考这个文档进行流水配置

Reviewer 2: ok

```python
from .common.train import train
...

train.dist.pipeline_parallel_size = 8
```

#### **Data Parallel + Tensor Parallel for 2D Parallel Training on 8 GPUs**

> In this example, 8 GPUs will be splitted into **2 groups**, each group contains **4 GPUs**, and the input data will be splitted into 2 parts by chunking in the batch dimension for data parallel training between 2 groups. And in each group, the weight of the layers in the model will be splited into 4 parts on 4 GPUs for tensor parallel training.

Reviewer 2: ```suggestion
In this example, 8 GPUs will be splitted into **2 groups**, each group contains **4 GPUs**, and the input data will be splitted into 2 parts by chunking in the batch dimension for data parallel training between 2 groups. The model is replicated between **2 data parellel groups**, within each group, the weight of each layers will be splited across **4 GPUs** for tensor parallel training.
```

Reviewer 2: 这样如何

Author: OK，我觉得可以

Author: 德澎你再看看还有没有其他表达不当的地方，我等等统一改

Author: 已改~

```python
from .common.train import train
...

train.dist.data_parallel_size = 8
```

#### **Pure Tensor Parallel Training on 8 GPUs**

In this example, the weight of the layers in the model be splitted into 8 parts for tensor parallel training on 8 GPUs.
```python
from .common.train import train
...

train.dist.tensor_parallel_size = 8
```

#### **Pure Pipeline Parallel Training on 8 GPUs**

In this example, 8 GPUs will be splitted into 8 stages, different layers of the model will be put on different stages automatically for pipeline parallel training.
```python
from .common.train import train
...

train.dist.pipeline_parallel_size = 8
```
**Note:** For models which have been configured with pipeline parallelism, e.g., BERT, GPT-2, T5 and VisionTransformer model in LiBai, you can simply update the distributed config to excute pipeline parallel training on these models. If you need to train your own model with pipeline parallel strategy, please refer to [Write Models]() for more details about configuring your own model with pipeline parallelism.

#### **Data Parallel + Tensor Parallel for 2D Parallel Training on 8 GPUs**

> In this example, 8 GPUs will be splitted into **2 groups**, each group contains **4 GPUs**, and the input data will be splitted into 2 parts by chunking in the batch dimension for data parallel training between 2 groups. And in each group, the weight of the layers in the model will be splited into 4 parts on 4 GPUs for tensor parallel training.

Reviewer 3: And in each group这句的splited要双写t

Author: 我改一下

```python
from .common.train import train
...

train.dist.data_parallel_size = 2
train.dist.tensor_parallel_size = 4
```

#### **Data Parallel + Pipeline Parallel for 2D Parallel Training on 8 GPUs**

> In this example, 8 GPUs will be splitted into **2 groups**, each group contains **4 GPUs**, and the input data will be splitted into 2 parts by chunking in the batch dimension for data parallel training. And each group contains **4 stages**, different layers in the model will be put on different stages automatically for pipeline parallel training.

Reviewer 2: 这里同样，最好首先要提一下模型在两个组之间是 replicated 的，然后再说在每个组内是 分 stage 的

Author: 可以的，我改一下

Author: 已改~

```python
from .common.train import train
...

train.dist.data_parallel_size = 2
train.dist.pipeline_parallel_size = 4
```

#### **Tensor Parallel + Pipeline Parallel for 2D Parallel Training on 8 GPUs**

> In this example, 8 GPUs will be splitted into **2 groups**, each group contains **4 GPUs**, and the weight of the layers in the model be splitted into 2 parts for tensor parallel training between 2 groups. And each group contains **4 stages**, different layers in the model will be put on different stages automatically for pipeline parallel training.

Reviewer 2: 张量+流水并行是这样么，2组4卡，每个组之间做张量并行，每个组内流水，是不是搞反了，每个组之间流水，组内张量并行

Author: 奥这里应该是写反了

```python
from .common.train import train
...

train.dist.tensor_parallel_size = 2
train.dist.pipeline_parallel_size = 4
```

#### **Data Parallel + Tensor Parallel + Pipeline Parallel for 3D Parallel Training on 8 GPUs**

> In this example, 8 GPUs will also be splitted into **2 groups**, but each group will also be splitted into **2 mini groups**, each mini groups contains 2 GPUs, the input data will be splitted into two parts by chunking in the batch dimension for data parallel training on 2 groups, and in each group, the weight of the layers in the model will be splitted into 2 parts for tensor parallel training on **2 mini groups**, and each mini group contrain **2 stages**, different layers in the model will be put on different stages for pipeline parallel training.

Reviewer 3: each mini-group（无s），'contrain'我猜想要表达的是'contains'

> In this example, 8 GPUs will also be splitted into **2 groups**, but each group will also be splitted into **2 mini groups**, each mini groups contains 2 GPUs, the input data will be splitted into two parts by chunking in the batch dimension for data parallel training on 2 groups, and in each group, the weight of the layers in the model will be splitted into 2 parts for tensor parallel training on **2 mini groups**, and each mini group contrain **2 stages**, different layers in the model will be put on different stages for pipeline parallel training.

Reviewer 2: 这里同上，张量和流水并行的顺序

```python
from .common.train import train
...

train.dist.data_parallel_size = 2
train.dist.tensor_parallel_size = 2
train.dist.pipeline_parallel_size = 2
```

**Note:** `train.dist.data_parallel_size` will be automatically calculated by `(gpu_nums / (tensor_parallel_size * pipeline_parallel_size))` if only `train.dist.tensor_parallel_size` and `train.dist.pipeline_parallel_size` are setted, for example:
```python
from .common.train import train
...
# only set tensor_parallel_size and pipeline_parallel_size
train.dist.tensor_parallel_size = 2
train.dist.pipeline_parallel_size = 2
```
And the `data_parallel_size` will be automatically setted to `(8 / (2 * 2)) = 2`

## Update Distributed Config with Command Line

> You can also use **command line** to control the parallelization mode as follows:

Reviewer 2: ```suggestion
You can also control the parallelization strategy by **command line** paremeters as follows:
```

Author: 已改~
